Integrate Polecat

tophmatthews opened the pull request:
Closes #20246

- [x] Make work with unit converter
- [x] Make work with substepping
- [x] Make work with inplace tests
- [x] Make work with new verification tests
- [x] Make work with advanced models (not in this PR)

Thanks to @aruybalid for the initial code improvements to include partitioning.

tophmatthews committed:
laromance_polecat_20246

tophmatthews commented:
Something is missing with the substepping still so those tests are failing, but everything else seems good to go with the tests in place. I still need to test this with the other ROMS, so it's not done yet.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/965015/) on 51631cf wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/965015/) on 51631cf wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/965015/) on 51631cf wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

cticenhour commented:
>The following file(s) are changed:
>
>  conda/libmesh/meta.yaml

@loganharbour is this notice going to show up every time there's a package change in next, even if the PR doesn't have an explicit change to this file in it? If so, is there a way we can limit the scope of this check to only the "actual" PR diff?

tophmatthews commented:
Whew! I was trying to figure out what I did wrong there!

loganharbour commented:
> > The following file(s) are changed:
> > conda/libmesh/meta.yaml
> 
> @loganharbour is this notice going to show up every time there's a package change in next, even if the PR doesn't have an explicit change to this file in it? If so, is there a way we can limit the scope of this check to only the "actual" PR diff?

This happens when someone rebases on next instead of devel and a conda change has not made it into devel. It's on my list to do git fu to probably get the merge base (which we use to determine the diffs), but it's not a priority because it doesn't happen often. I would suggest instead that people just rebase off of devel like we ask them.

tophmatthews committed:
Add partition base class
Closes #20246

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/965239/) on 3222fa3 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/965239/) on 3222fa3 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/965239/) on 3222fa3 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

tophmatthews committed:
Redid verification tests for AD and nonAD

tophmatthews committed:
update gold files for substepping
Due to changes in the algorithm, it seems as though the nonlinear 
convergence is improved considerablly. In the previous test for 
substepping, the timestep was cut due to too many nonlinear iterations. 
In the current implementation, the timestep is not limited the same way, 
resulting in different (fewer) timesteps. These tests still show that 
substepping with large timesteps does help get closer to the test with 
many small timesteps, so I think these should still hold valid.

tophmatthews commented:
@recuero This new version of LaRomance has improved nonlinear convergence. Consequently, the tests you made to ensure that substepping was helping with the total error are not the same, and require a re-gold. The output shows that substepping with large timesteps approaches the value of no-substepping with small timesteps...I think that should be good enough for the tests? Can you verify I'm not missing something?

tophmatthews commented on modules/tensor_mechanics/include/materials/LAROMANCEPartitionStressUpdateBase.h:
```diff
@@ -0,0 +1,55 @@
+  virtual void computeDSecondPartitionWeightDStress(
+      GenericReal<is_ad> & dsecond_partition_weight_dstress) override;
+
+  /// todo
```
I have to update this doco string still...

recuero commented:
> the output shows that substepping with large timesteps approaches the value of no-substepping with small timesteps

Yeah, that's the main idea. I went through the updated csv files in this PR and the results make qualitatively sense to me. Usage of substepping still shows improved accuracy over regular stepping. The differences due to the updated stepping (`dt` changes) do not seem to affect the results a lot.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/965455/) on e0eaaa7 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>016200</th>
      <th colspan=4>#20386 e0eaaa</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.44%</td>
      <td>88.98%</td>
      <td>-0.46%</td>
      <td>51.94%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26169</td>
      <td>26268</td>
      <td>+99</td>
      <td>241</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3090</td>
      <td>3254</td>
      <td>+164</td>
      <td>223</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20386/coverage_diff/tensor_mechanics.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20386/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20386/coverage/xfem)

</details>

## Warnings

- `tensor_mechanics` new line coverage rate 51.94% is less than the suggested 90.0%



This comment will be updated on new commits.

tophmatthews committed:
doco update

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/965908/) on aae8843 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20386/site)

This comment will be updated on new commits.

tophmatthews commented:
@dschwen @bwspenc @lynnmunday Here is the Polecat integration we've been talking about for some time. It seems as though convergence is improved, although I'm not sure where or how...likely just a consequence of cleanups. Please run it with your current models (as we are) and ensure that similar results are achieved, and review the changes here. Thanks!

tophmatthews commented:
But yes, it's just a swap. I sent you better verification tests and data though, so you'll want to use what I emailed you to build and test the proprietary model.

tophmatthews commented:
> @tophathews,

Coffee still waiting to kick in?

dschwen commented:
> > @tophathews,
> 
> Coffee still waiting to kick in?

🎩 ⛏️  That's not your name?

tophmatthews commented:
Oh, maybe zip problems...I'll a different way.

tophmatthews committed:
Enable substepping in all laromance models

tophmatthews commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -165,25 +218,49 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::LAROMANCEStressUpdateBaseTempl(
+bool
+LAROMANCEStressUpdateBaseTempl<is_ad>::substeppingCapabilityEnabled()
+{
+  return this->template getParam<bool>("use_substep");
```
FYI @dschwen this is required (return true led to errors?)

dschwen replied:
Got it. That also means the function name is good. Not how I would have designed it though, as the parameter is added by a parent class and I prefer parameter handling to happen at the same level.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/967548/) on 726d0e8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20386/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20386/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9d27e0165cc0ab39b6b705a812b46234be3382cd`

tophmatthews commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -879,40 +1188,49 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::computeStressFinalize(
-  for (unsigned int t = 0; t < _num_tiles; ++t)
-    if (_weights[t])
-      _wall_dislocation_increment += _weights[t] * computeROM(t, _wall_output_index);
+  if (_input_values[_stress_input_index])
```
This fixed up the `nan`'s I was seeing @dschwen , not sure if it was the same for you or similar to your fix?

tophmatthews commented:
Hey @dschwen and @bwspenc , there are a couple more things I need to do to stamp this integration as done, but I think it'll be easier to get this in. @dschwen already showed that answers are the same as previous implementations, but convergence has a bit of problem in certain inputs. There are some fixes there that we are working on, but I don't want to get this tangled any more. Can we merge this?

dschwen commented:
Yeah, that's fine with me

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/996079/) on cdef0f6 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 93.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 86.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 91.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 83.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_conduction` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 88.0%)
- Failed to generate `misc` coverage rate (required: 55.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `peridynamics` coverage rate (required: 78.0%)
- Failed to generate `phase_field` coverage rate (required: 89.0%)
- Failed to generate `porous_flow` coverage rate (required: 96.0%)
- Failed to generate `ray_tracing` coverage rate (required: 94.0%)
- Failed to generate `rdg` coverage rate (required: 65.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 86.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

tophmatthews commented:
Actually...hold off on the merge, I found the bug @dschwen ... I think 🤞

tophmatthews committed:
Fix derivative bug

tophmatthews committed:
Fix global limit calculation. It's still not perfect, and mising tiles at the extremes may still hit errors, but it should be good for most applications.

tophmatthews commented:
Alright, give it a try @dschwen

tophmatthews commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -789,7 +789,7 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::computeResidual(
           total_rom_effective_strain_inc += _partition_weights[p] * _weights[p][t] * rom;
 
-          dtotal_rom_effective_strain_inc_dstress =
+          dtotal_rom_effective_strain_inc_dstress +=
```
🙄 
(In case you were curious @dschwen )

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/998386/) on e0d62f9 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

dschwen commented:
Running this right now and it looks pretty good! Let me do some more checks with cranked up loading.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/998424/) on 0fdb42e wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/998424/) on 0fdb42e wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/998424/) on 0fdb42e wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

dschwen commented on modules/tensor_mechanics/src/materials/LAROMANCEPartitionStressUpdateBase.C:
```diff
@@ -0,0 +1,124 @@
+  // _old_strain_input_index)
+  std::vector<GenericReal<is_ad>> scaled_input_values(_input_values.size() - 1);
+  unsigned int inc = 0;
+  for (unsigned int i = 0; i < _input_values.size(); i++)
```
We tend to use

```suggestion
  for (const auto i: index_range(_input_values))
```

in new code these days. (`#include "libmesh/int_range.h"`)

dschwen replied:
Sure we do that plenty. Here you seem to need the index though.

tophmatthews replied:
Yeah, totally. Just curious. Is this a fix it now or fix it later suggestion here?

dschwen replied:
Up to you. I can approve this now.

dschwen reviewed (commented):
Works perfectly now!

lynnmunday commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -408,144 +531,210 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::computeStressInitialize(
+              weights[p][t] *= (1.0 - sigmoid(0.0, _input_limits[p][t][in_index][0], input));
+            input = _input_limits[p][t][in_index][0];
+          }
+          else if (_window_failure[in_index][0] == WindowFailure::USELIMIT)
```
should this set weights=0?  Same with line 670.  Is USELIMIT the same as DONOTHING enum you have commented out?

tophmatthews replied:
If USELIMIT is used, it just changes the input limit to the edge of the box of applicability, and runs it normally. DONOTHING sets weight to zero.

lynnmunday replied:
what does USELIMIT set the weights to?  I don't see where weights would be set.

tophmatthews replied:
Essentially multiplies by 1. You can't set the weights to anything, except zero, only multiple them by the local weight, since it's the accumulation of weights for all input values.

tophmatthews committed:
Turn some things into pairs, and fix compile warnings for bool *

dschwen commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -328,12 +325,12 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::initialSetup()
   {
-    _global_limits[i][0] = std::numeric_limits<Real>::max();
-    _global_limits[i][1] = 0.0;
+    _global_limits[i].first = std::numeric_limits<Real>::max();
+    _global_limits[i].second = 0.0;
```
```suggestion
    _global_limits[i]= {std::numeric_limits<Real>::max(), 0.0};
```
?

tophmatthews replied:
Yes, that's nicer...next round?

dschwen commented:
FYI

|phase|NEML    | old LaRomance | new LaRomance |
|--------|-----------|-----------------------|------------------------|
|1|2072s|5042s|8281s|
|2|1331s|7787s|24612s|

tophmatthews commented:
There is one more fix that @aruybalid has queued up that may help that @dschwen . Can we get this in so he can apply some new patches? After a good review from you and @lynnmunday of course.

lynnmunday commented:
@tophmatthews I approve.  I don't think I have an approval button I can press though.

lynnmunday commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -10,53 +10,91 @@
+  params.addParam<MooseEnum>(
+      "environment_input_window_high_failure",
+      extrapolated_upper_limit_behavior,
       "What to do if environmental factor is outside the global window of applicability.");
```
you can fix that when you make your next patch.

tophmatthews commented:
Oh wow...Profiling pointed out some dummy mistakes...I sped it up 90% on a dummy 1 element problem...standby

tophmatthews committed:
Preallocate containers for massive speed ups, yahoo!

bwspenc reviewed (commented):
I looked through this, and the code looks clean. Just one question, though: based on the naming of the base classes, I would have expected everything related to the partitioned model to be in LAROMANCEPartitionStressUpdateBase, but there's an awful lot of partition-related code in LAROMANCEStressUpdateBase. Just curious what the thinking is there.

tophmatthews commented:
> but there's an awful lot of partition-related code in LAROMANCEStressUpdateBase

It's actually not a bad point that the `computePartitionWeights` should just return `1` in `LAROMANCEStressUpdateBase`, and be overriden in `LAROMANCEPartitionStressUpdateBase`. That way `computeSecondPartitionWeight` and `computeDSecondPartitionWeightDStress` could only be defined in `LAROMANCEPartitionStressUpdateBase` as well.

As for the other parts, most of it is the required extra dimension on the vectors, so I think that needs to stay. I clean up those pieces though, thanks for the tip @bwspenc

aruybalid commented:
Hi guys, sorry to be a bit of noob commenting in Github, so I am not sure if this is the correct way and place to provide comments/feedback. I was running some tests and ran into the error:
```
*** ERROR ***
In rom_stress_prediction_base: input parameter with value (5.2476e+12) is out of upper global range (2.78853e+13)
```
which is weird, because that value is not the upper limit. I think a copy/paste error slipped in the `computeTileWeight` method in `LAROMANCEStressUpdateBase.C`:

```
      // If input is above the upper tile limit
      else if (input > _input_limits[p][t][in_index][1])
      {
        if (_input_limits[p][t][in_index][1] == _global_limits[in_index].second)
        {
          if (_window_failure[in_index].second == WindowFailure::EXTRAPOLATE)
            mooseError("Internal error. Extrapolate not appropriate for upper bound");
          else if (_window_failure[in_index].second == WindowFailure::USELIMIT)
            input = _input_limits[p][t][in_index][1];
          else
          {
            // Set input to edge of limit so it is found later in computePartitionWeights
            input = _input_limits[p][t][in_index][0];
            weights[p][t] = 0.0;
            std::stringstream msg;
            msg << "In " << _name << ": input parameter with value ("
                << MetaPhysicL::raw_value(input) << ") is out of upper global range ("
                << _input_limits[p][t][in_index][1] << ")";

            if (_window_failure[in_index].first == WindowFailure::WARN)
              mooseWarning(msg.str());
            else if (_window_failure[in_index].first == WindowFailure::ERROR)
              mooseError(msg.str());
            else if (_window_failure[in_index].first == WindowFailure::EXCEPTION)
              mooseException(msg.str());

```
I believe it should be:
```
      // If input is above the upper tile limit
      else if (input > _input_limits[p][t][in_index][1])
      {
        if (_input_limits[p][t][in_index][1] == _global_limits[in_index].second)
        {
          if (_window_failure[in_index].second == WindowFailure::EXTRAPOLATE)
            mooseError("Internal error. Extrapolate not appropriate for upper bound");
          else if (_window_failure[in_index].second == WindowFailure::USELIMIT)
            input = _input_limits[p][t][in_index][1];
          else
          {
            // Set input to edge of limit so it is found later in computePartitionWeights
            input = _input_limits[p][t][in_index][1];
            weights[p][t] = 0.0;
            std::stringstream msg;
            msg << "In " << _name << ": input parameter with value ("
                << MetaPhysicL::raw_value(input) << ") is out of upper global range ("
                << _input_limits[p][t][in_index][1] << ")";

            if (_window_failure[in_index].second == WindowFailure::WARN)
              mooseWarning(msg.str());
            else if (_window_failure[in_index].second == WindowFailure::ERROR)
              mooseError(msg.str());
            else if (_window_failure[in_index].second == WindowFailure::EXCEPTION)
              mooseException(msg.str());
```

tophmatthews commented:
Nice catch! Next time just comment directly in the spot in the code so I don't have to look through 30 lines to spot-the-difference 😄

tophmatthews committed:
Moved things to partition base and fix copy-paste error

aruybalid commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -408,144 +526,213 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::computeStressInitialize(
+          else
+          {
+            // Set input to edge of limit so it is found later in computePartitionWeights
+            input = _input_limits[p][t][in_index][0];
```
`input = _input_limits[p][t][in_index][1];`

tophmatthews replied:
🤘

aruybalid commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -408,144 +526,213 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::computeStressInitialize(
+                << MetaPhysicL::raw_value(input) << ") is out of upper global range ("
+                << _input_limits[p][t][in_index][1] << ")";
+
+            if (_window_failure[in_index].first == WindowFailure::WARN)
```
`if (_window_failure[in_index].second == WindowFailure::WARN)`

aruybalid commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -408,144 +526,213 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::computeStressInitialize(
+
+            if (_window_failure[in_index].first == WindowFailure::WARN)
+              mooseWarning(msg.str());
+            else if (_window_failure[in_index].first == WindowFailure::ERROR)
```
`else if (_window_failure[in_index].second == WindowFailure::ERROR)`

tophmatthews replied:
Oh I see there was a lot...

aruybalid commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -408,144 +526,213 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::computeStressInitialize(
+              mooseWarning(msg.str());
+            else if (_window_failure[in_index].first == WindowFailure::ERROR)
+              mooseError(msg.str());
+            else if (_window_failure[in_index].first == WindowFailure::EXCEPTION)
```
`else if (_window_failure[in_index].first == WindowFailure::EXCEPTION)`

tophmatthews commented on modules/tensor_mechanics/include/materials/LAROMANCEPartitionStressUpdateBase.h:
```diff
@@ -0,0 +1,92 @@
+  ///@}
+
+  ///@{ Method and container for the inducing points of the Gaussian Process Regression model
+  virtual DenseVector<Real> getClassificationVind() = 0;
```
This will require redefinition of this method in the derivative classes.

aruybalid commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -668,18 +668,18 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::computeTileWeight(
             weights[p][t] = 0.0;
             std::stringstream msg;
             msg << "In " << _name << ": input parameter with value ("
                 << MetaPhysicL::raw_value(input) << ") is out of upper global range ("
```
print the input before it was reset to limit:
```
msg << "In " << _name << ": input parameter with value ("
                << input_before_set_to_limit << ") is out of upper global range ("
                << _input_limits[p][t][in_index][1] << ")";
```

tophmatthews replied:
Good catch!

aruybalid replied:
Do the same for the lower limit: lines 643 and 639 (for some reason I don't get the little + symbol to add a comment in those lines?)

aruybalid commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -668,18 +668,18 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::computeTileWeight(
           {
             // Set input to edge of limit so it is found later in computePartitionWeights
-            input = _input_limits[p][t][in_index][0];
+            input = _input_limits[p][t][in_index][1];
```
store value of input before setting it to limit value (for later console print message):
```
GenericReal<is_ad> input_before_set_to_limit = input;
input = _input_limits[p][t][in_index][1];
```

tophmatthews replied:
No, I don't think I need that, I can just do `input = _input_limits[p][t][in_index][1]` after I make the msg, no?

aruybalid replied:
jep, that works. Also in the lower limit case..

tophmatthews commented:
We're in (what feels like) a narrow window where moose/next is green so this test is passing, is this good to go in @dschwen @bwspenc ?
